bridgenode/flatfileworker.go

```diff
@@ -0,0 +1,200 @@
+
+	// Grab either proof bytes and write em to offset / proof file, OR, get a TTL result
+	// and write that.  Will this lock up if it keeps doing proofs and ignores ttls?
+	// it should keep both buffers about even.  If it keeps doing proofs and the ttl
+	// buffer fills, then eventually it'll block...?
+	// Also, is it OK to have 2 different workers here?  It probably is, with the
+	// ttl side having read access to the proof writing side's last written proof.
+	// then the TTL side can do concurrent writes.  Also the TTL writes might be
+	// slow since they're all over the place.  Also the offsets should definitely
+	// be in ram since they'll be accessed a lot.
+
+	// TODO ^^^^^^ all that stuff.
+	for {
+		pbytes := <-proofChan
+		// write to offset file first
+		_, _ = offsetFile.Seek(int64(proofWriteHeight)*8, 0)
+		err = binary.Write(offsetFile, binary.BigEndian, curOffset)
+		if err != nil {
+			fmt.Printf(err.Error())
+			return
+		}
+
+		// write to proof file
+		// first write big endian proof size uint32 (proof never more than 4GB)
+		err = binary.Write(proofFile, binary.BigEndian, uint32(len(pbytes)))
+		if err != nil {
+			fmt.Printf(err.Error())
+			return
+		}
+
+		// then write the proof
+		written, err := proofFile.Write(pbytes)
+		if err != nil {
+			fmt.Printf(err.Error())
+			return
+		}
+		curOffset += int64(written) + 4 // the size comes first
+		proofWriteHeight++
+
+		// send offset to the ttl worker after proofs are written to disk
+		offsetChan <- curOffset
+		fileWait.Done()
+		// fmt.Printf("flatFileBlockWorker h %d done\n", proofWriteHeight)
+	}
+
+}
+
+func flatFileTTLWorker(
+	ttlResultChan chan ttlResultBlock,
+	offsetChan chan int64,
+	fileWait *sync.WaitGroup) {
+
+	var inRamOffsets []int64
+
+	// the offset for block 0 is 0?  kindof weird
+	inRamOffsets = append(inRamOffsets, 0)
+
+	maxOffsetHeight := int32(1) // start at 1?
+
+	// for the pFile
+	proofFile, err := os.OpenFile(
+		util.PFilePath, os.O_APPEND|os.O_CREATE|os.O_WRONLY, 0600)
+	if err != nil {
+		panic(err)
+	}
+
+	for {
+		// get offset data; if there is none, try getting ttl data
+		select {
+		case nextOffset := <-offsetChan:
+			// got an offset, expand in ram offsets and write
+			inRamOffsets = append(inRamOffsets, nextOffset)
+			maxOffsetHeight++
+
+		case ttlRes := <-ttlResultChan:
+			// fmt.Printf("got ttlres h %d\n", ttlRes.Height)
+			for ttlRes.Height > maxOffsetHeight {
+				// we got a ttl result before the offset.  We need the offset data
+				// first, so keep reading offsets until we're caught up
+				inRamOffsets = append(inRamOffsets, <-offsetChan)
+				maxOffsetHeight++
+			}
+			// for all the TTLs, seek and overwrite the empty values there
+			for _, c := range ttlRes.Created {
+				// seek to the location of that txo's ttl value in the proof file
+
+				// fmt.Printf("want ram offset %d, only have up to %d\n",
+				// 	c.createHeight, len(inRamOffsets))
+
+				_, _ = proofFile.Seek(
```

Comment:
i would not be surprised if this slows things down quite a bit...
keeping the proofs&ttls for the last n blocks in memory could help but lets see if this is actually a problem first.